src/algebra/star/basic.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison.
+-/
+import tactic.apply_fun
+import algebra.ordered_ring
+import algebra.big_operators.basic
+
+/-!
+# star_monoid, star_ring
+
+We introduce the basic algebraic notions of star monoids, and star rings.
+Star algebras are introduced in `algebra.algebra.star`.
+
+These are implemented as "mixin" typeclasses, so to summon a star ring (for example)
+one needs to write `(R : Type) [ring R] [star_ring R]`.
+This avoids difficulties with diamond inheritance.
+
+For now we simply do not introduce notations,
+as different users are expected to feel strongly about the relative merits of
+`r^*`, `r†`, `rᘁ`, and so on.
+
+Our star rings are actually star semirings, but of course we can prove
+`star_neg : star (-r) = - star r` when the underlying semiring is a ring.
+-/
+
+
+universes u v
+
+/--
+Notation typeclass (with no default notation!) for an algebraic structure with a star operation.
+-/
+class has_star (R : Type u) :=
+(star : R → R)
+
+variables {R : Type u}
+
+/--
+A star operation (e.g. complex conjugate).
+-/
+def star [has_star R] (r : R) : R := has_star.star r
+
+/--
+Typeclass for a star operation with is involutive.
+-/
+class star_involutive (R : Type u) extends has_star R :=
+(star_star : ∀ r : R, star (star r) = r)
+
+@[simp] lemma star_star [star_involutive R] (r : R) : star (star r) = r :=
+star_involutive.star_star _
+
+lemma star_injective [star_involutive R] : function.injective (star : R → R) :=
+begin
```

Comment:
This can be a oneliner, if we use `involutive` as suggested above.

Reply:
It seems odd to me for `star_involutive` to be the typeclass but `star_injective` to be an `injective` instance. Can you think of a better name for the typeclass so that we can have `star_involutive : function.involutive (star : R → R)`? Perhaps `has_involutive_star`?

Reply:
I went with `has_involutive_star`. The long name is unlikely to matter, since likely only `star_ring` will be used anyway.